[macOS] Screen/window picker sometimes remains after capture stops

eric-carlson opened the pull request:
#### c44cb30770746f9750c34f9444895d1eaa549758
<pre>
[macOS] Screen/window picker sometimes remains after capture stops
<a href="https://bugs.webkit.org/show_bug.cgi?id=259189">https://bugs.webkit.org/show_bug.cgi?id=259189</a>
rdar://112195888

Reviewed by Youenn Fablet and Jer Noble.

Deactivate the screen/window picker during session cleanup unless another picking session
is active, so the menu is hidden when there are no active capture streams.

* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(WebCore::ScreenCaptureKitCaptureSource::stop): Clear `m_sessionSource` immediately instead
of in the stop lambda, there is no reason to wait.

* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.h
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.mm:
(WebCore::ScreenCaptureKitSharingSessionManager::cleanupSessionSource): If a prompt isn&apos;t
active, call `cancelPicking` to cleanup.

Canonical link: <a href="https://commits.webkit.org/266041@main">https://commits.webkit.org/266041@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d737b211bd27cd5dc6c24bf98a12b851e5538683

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12654 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13298 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14393 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12113 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12715 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15485 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13000 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14813 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12818 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13590 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10714 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14835 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10867 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11459 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18540 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11952 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11625 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14822 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12098 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10013 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11346 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3105 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15679 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11955 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.mm:
```diff
@@ -542,11 +542,13 @@ - (void)stopObservingPicker:(SCContentSharingPicker *)picker
 
-    auto sharingSession = activeSource->sharingSession();
+    if (!promptingInProcess())
+        cancelPicking();
```
Should we do this check even if there is no activeSource?

jernoble replied:
Looks like we should. If there are no active sources, we do even more teardown in cancelPicking().

eric-carlson replied:
Good idea, fixed

jernoble commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
```diff
@@ -183,15 +183,14 @@ - (void)stream:(SCStream *)stream didOutputSampleBuffer:(CMSampleBufferRef)sampl
-            weakThis->m_sessionSource = nullptr;
-            if (error)
-                weakThis->sessionFailedWithError(WTFMove(error), "-[SCStream stopCaptureWithCompletionHandler:] failed"_s);
+            weakThis->sessionFailedWithError(WTFMove(error), "-[SCStream stopCaptureWithCompletionHandler:] failed"_s);
```
As long as you're in here, this could now be:
```
if (weakThis && error)
    weakThis->sessionFailedWithError(WTFMove(error), "-[SCStream stopCaptureWithCompletionHandler:] failed"_s);
```

eric-carlson replied:
Fixed

jernoble reviewed (approved):
Needs a change to the commit message to remove the `OOPS!`.

eric-carlson committed:
[macOS] Screen/window picker sometimes remains after capture stops
https://bugs.webkit.org/show_bug.cgi?id=259189
rdar://112195888

Reviewed by Youenn Fablet and Jer Noble.

Deactivate the screen/window picker during session cleanup unless another picking session
is active, so the menu is hidden when there are no active capture streams.

* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(WebCore::ScreenCaptureKitCaptureSource::stop): Clear `m_sessionSource` immediately instead
of in the stop lambda, there is no reason to wait.

* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.h
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.mm:
(WebCore::ScreenCaptureKitSharingSessionManager::cleanupSessionSource): If a prompt isn't
active, call `cancelPicking` to cleanup.

Canonical link: https://commits.webkit.org/266041@main
